src/tests.c

```diff
@@ -0,0 +1,224 @@
+#include <openssl/bio.h>
+#include <openssl/cms.h>
+#include <openssl/err.h>
+#include <openssl/pem.h>
+#include <openssl/safestack.h>
+#include <openssl/x509.h>
+#include <openssl/x509_vfy.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/time.h>
+
+#include "./verify.h"
+
+#ifndef BENCH_ITER
+#define BENCH_ITER 100
+#endif
+
+// Set up some stdio wrappers for dumping BIO messages
+BIO *bio_out, *bio_err;
+
+// Returns the number of bytes read, and sets the contents ** to the start
+// of the memory buffer
+return_t read_complete_file(char *filename, char **value, long *length) {
+  FILE *f = fopen(filename, "r");
+  if (!f) {
+    perror("opening file");
+    return VF_FAIL;
+  }
+
+  if (-1 == fseek(f, 0, SEEK_END)) {
+    perror("seeking to file end");
+    fclose(f);
+    return VF_FAIL;
+  }
+
+  long size = ftell(f);
+  *length = size;
+
+  if (-1 == fseek(f, 0, SEEK_SET)) {
+    perror("seeking to beginning of file");
+    fclose(f);
+    return VF_FAIL;
+  }
+
+  *value = malloc(size + 1);
+  if (*value == NULL) {
+    perror("allocating buffer");
+    fclose(f);
+    return VF_FAIL;
+  }
+
+  int read = fread(*value, 1, size, f);
+
+  if (!ferror(f) || read != size) {
+    fclose(f);
+    return VF_SUCCESS;
+  } else {
+    perror("Reading file");
+    fclose(f);
+    return VF_FAIL;
+  }
+}
+
+// Print out a complete BIO to console, probably only safe to do on simple BIO
+// types, like memory buffer backed ones
+//
+// NOTE: This code is only intended for testing/debugging
+void print_bio_s_mem(BIO *bio, size_t len) {
+  char *value = malloc(sizeof(char) * len);
+  BIO_read(bio, value, len);
+  BIO_write(bio_out, value, len);
+  BIO_printf(bio_out, "\n");
+  free(value);
+}
+
+char *memdup(char *src, long len) {
+  char *x = malloc(len);
+  memcpy(x, src, len);
+  return x;
+}
+
+// Do some tests.  I don't care about leaking memory in this function
```

Comment:
It can be a little more welcoming to use "we" in comments instead of 'I", or just avoid a pronoun altogether -- otherwise a reader looks at this code and thinks, "Oh, this belongs to someone, I shouldn't touch it".  So perhaps "Memory leaks in this function are not important, since this is just a test harness, but covering the tested code is very important".

Reply:
Most removed, left in testing code where it's difficult to reword